<details>
<summary><b>Pre-PR validation gate</b> (click to expand)</summary>

| # | Question | Answer |
|---|----------|--------|
| a | **Deterministic repro?** | Yes. Injected an 8s sleep at the start of `runSnapshotMaintenance` (the Phase-4 move step) and ran the unmodified test: 3/3 FAIL with `AssertionError: assert 2 == 1` (exact CI signature). The clean build is too fast to hit the window reliably (15/15 passed), confirming this is a timing race, not a logic error. |
| b | **Root cause explained?** | The `Created persistent snapshot {idx}` log line is emitted in the under-lock publish step (`KeeperStateMachine::publishWrittenSnapshot`), but the previous non-latest snapshot is removed from the local "latest" disk only in the deferred, outside-the-lock step (`runSnapshotMaintenance` -> `moveSnapshotCandidate`), which runs after the log line. `selectSnapshotsToMove` additionally defers a move while the old snapshot is transiently pinned (`use_count > 1`). The test waited on the log line, then checked the local count once, so it raced the still-pending move and saw 2 files. |
| c | **Fix matches root cause?** | Yes. The relocation is eventually consistent, so the test polls for the local count to settle to 1 (up to 60s) instead of asserting once. No bound widening, no `no-random-*` tag, no data reduction. |
| d | **Test intent preserved / new tests added?** | Yes. The test still asserts exactly one local snapshot remains, that it is the latest, and that all others moved to S3. Only the single-shot check becomes a poll. Mirrors the existing `assert_single_local_log` helper used by the sibling `test_logs_with_disks`. No source code changed, so no new regression test is needed. |
| e | **Both directions demonstrated?** | Yes. 8s-delay build: bare assertion 3/3 FAIL, polling fix 3/3 PASS. Clean build: polling fix 10/10 PASS. |
| f | **Fix is general across code paths?** | Yes. Both bare `len == 1` assertions in `test_snapshots_with_disks` (after each `csnp`) share this race and are both converted to the poll. The third site (`assert set(...) == set(...)` after the final S3->local move) is not racy: that relocation happens during the server-restart load path, which completes before `wait_until_connected` returns, so it is left unchanged. |
| g | **Fix generalizes across inputs?** | N/A (test-only flaky fix, not a code bug). |
| h | **Backward compatible?** | N/A (test-only flaky fix, no server behavior or settings changed). |
| i | **Invariants and contracts preserved?** | N/A (test-only flaky fix, no source invariants touched). |

</details>

Session id: cron:clickhouse-worker-slot-1:20260625-150000

--- a/tests/integration/test_keeper_disks/test.py
+++ b/tests/integration/test_keeper_disks/test.py
@@ -59,7 +59,8 @@ def stop_clickhouse(cluster, node, cleanup_disks):
     node.exec_in_container(["rm", "-rf", "/var/lib/clickhouse/coordination/logs"])
     node.exec_in_container(["rm", "-rf", "/var/lib/clickhouse/coordination/snapshots"])
 
-    s3_objects = list_s3_objects(cluster, prefix="")
+    # Wipe everything, including any leftover "tmp_" markers, so the bucket is clean.
+    s3_objects = list_s3_objects(cluster, prefix="", exclude_tmp=False)
     if len(s3_objects) == 0:
         return
 
@@ -106,7 +107,15 @@ def setup_local_storage(cluster, node):
     )
 
 
-def list_s3_objects(cluster, prefix=""):
+# Keeper writes an empty "tmp_<name>" marker on the target disk while moving a snapshot/log
+# file between disks and removes it once the copy completes (moveFileBetweenDisks in
+# KeeperCommon.cpp). Such markers are never real snapshots/logs: the server skips
+# "tmp_"-prefixed entries when scanning disks. A listing taken while a move is in flight can
+# transiently observe a marker, so exclude it instead of counting it as a snapshot/log file.
+tmp_keeper_file_prefix = "tmp_"
+
+
+def list_s3_objects(cluster, prefix="", exclude_tmp=True):
     minio = cluster.minio_client
     prefix_len = len(prefix)
     return [
@@ -114,11 +123,16 @@ def list_s3_objects(cluster, prefix=""):
         for obj in minio.list_objects(
             cluster.minio_bucket, prefix=prefix, recursive=True
         )
+        if not (
+            exclude_tmp
+            and os.path.basename(obj.object_name).startswith(tmp_keeper_file_prefix)
+        )
     ]
 
 
 def get_local_files(path, node):
     files = node.exec_in_container(["ls", path]).strip().split("\n")
+    files = [f for f in files if not f.startswith(tmp_keeper_file_prefix)]
     files.sort()
     return files
 

